src/sparseml/pytorch/yolov8/trainers.py

```diff
@@ -58,17 +114,117 @@ def train(self):
+            self.args.epochs = self.manager.max_epochs
+
+            if self.manager.learning_rate_modifiers:
+                self.scheduler = _NullLRScheduler()
+
+            self.manager.initialize(
+                self.model, epoch=self.start_epoch, loggers=self.logger_manager
+            )
+
+            # NOTE: we intentionally don't divide number of batches by gradient
+            # accumulation.
+            # This is because yolov8 changes size of gradient accumulation during
+            # warmup epochs, which is incompatible with SparseML managers
+            # because they assume a static steps_per_epoch.
+            # Instead, the manager will effectively ignore gradient accumulation,
+            # and we will call self.scaler.emulated_step() if the batch was
+            # accumulated.
+            self.steps_per_epoch = len(self.train_loader)  # / self.accumulate
+
+            self.scaler = self.manager.modify(
+                self.model,
+                self.optimizer,
+                steps_per_epoch=self.steps_per_epoch,
+                epoch=self.start_epoch,
+                wrap_optim=self.scaler,
+            )
+
+    def callback_on_train_epoch_start(self):
+        # NOTE: this callback is registered in __init__
+        if self.manager is not None and self.manager.qat_active(epoch=self.epoch):
+            if self.scaler is not None:
+                self.scaler._enabled = False
+            self.ema = None
+
+        self.epoch_step = 0
+
+    def callback_on_train_batch_start(self):
+        self.do_emulated_step = True
 
     def optimizer_step(self):
-        return super().optimizer_step()
+        super().optimizer_step()
+        self.do_emulated_step = False
+
+    def callback_on_train_batch_end(self):
+        # Here is where we handle the changing gradient accumulation values by doing
+        # an emulated step if we accumulated gradients of this batch
+        if self.do_emulated_step:
+            self.scaler.emulated_step()
+
+        step = self.epoch * self.steps_per_epoch + self.epoch_step
+        for key, value in self.label_loss_items(self.tloss).items():
+            self.logger_manager.log_scalar(key, value, step=step)
 
     def save_model(self):
+        # NOTE: identical to super().save_model() with the addition of recipe key
+        # in the checkpoint
+
+        epoch = -1 if self.epoch == self.epochs - 1 else self.epoch
+
+        if self.checkpoint_manager is not None:
+            if epoch >= 0:
+                epoch += self.checkpoint_manager.max_epochs
+            manager = ScheduledModifierManager.compose_staged(
+                self.checkpoint_manager, self.manager
+            )
+        else:
+            manager = self.manager if self.manager is not None else None
+
         ckpt = {
-            "epoch": self.epoch,
+            "epoch": epoch,
             "best_fitness": self.best_fitness,
             "model": deepcopy(de_parallel(self.model)).half(),
```

Comment:
Should probably be updated to save as weight dict, to avoid the same pickling issues as in yolov5

Reply:
Oooh good call. I think we also expect `"state_dict"` key instead of `"model"` right? At least for torchvision we did... Does yolov5 use model or state_dict?

Reply:
it's still "model" in `yolov5`. I think it'd be fine to keep it that way in yolov8 as well. Also note same should apply to ema. ema will also need some conditional logic for saving when it's `None` during quantizations

Reply:
Good call, will fix that!